proposals/p2070.md

```diff
@@ -0,0 +1,137 @@
+established how member name lookup was performed in a number of situations,
+including for template parameters. Generally speaking, the rule can be
+summarized as:
+
+> Look up the given name in all the relevant types and interfaces. If a unique
+> name name is found, use it. If more than one name is found, require the user
+> to add qualification to disambiguate.
+
+Applying this rule to `where` constraints to remove the semantic difference
+between the two forms was discussed:
+
+-   [in the #syntax on Discord](https://discord.com/channels/655572317891461132/709488742942900284/979811441852960808),
+    and
+-   [during the 2022-06-01 weekly sync meeting](https://docs.google.com/document/d/1dwS2sJ8tsN3LwxqmZSv9OvqutYhP71dK9Dmr1IXQFTs/edit?resourcekey=0-NxBWgL9h05yD2GOR3wUisg#heading=h.qarzfirrcrgf).
+
+## Proposal
+
+The proposal is to use the single `==` syntax, which is more consistent with
+other `where` constraint clauses. This syntax would consistently use the same
+semantics whether the constraint uses a concrete type or not. In both cases,
+when looking up a name in a constrained type variable, the nae will be looked up
+using the information from both sides of the `==`. As usual, name conflicts are
+resolved by explicit disambiguation.
+
+## Details
+
+The new semantics are described in the
+["Satisfying both type-of-types" section](/docs/design/generics/details.md#satisfying-both-type-of-types)
+of the
+[Generics: Details document at `docs/design/generics/details.md`](/docs/design/generics/details.md).
+In addition, references to the old syntax have been updated in:
+
+-   [`docs/design/expressions/arithmetic.md`](/docs/design/expressions/arithmetic.md)
+-   [`docs/design/expressions/bitwise.md`](/docs/design/expressions/bitwise.md)
+-   [`docs/design/expressions/member_access.md`](/docs/design/expressions/member_access.md)
+-   [`docs/design/generics/details.md`](/docs/design/generics/details.md)
+-   [`docs/design/generics/overview.md`](/docs/design/generics/overview.md)
+-   [`docs/design/generics/terminology.md`](/docs/design/generics/terminology.md)
+
+## Rationale
+
+The new semantics are more consistent with the rest of Carbon, and is simpler.
+This directly helps the
+["code that is easy to read, understand, and write"](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+goal. In addition, fewer syntactic and semantics constructs, making it easier to
+develop
+[language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem).
+
+## Alternatives considered
+
+Mainly this proposal is being evaluated against the status quo of the existing
+design.
+
+### `=` instead of `==`
+
+We did consider which of `=` and `==` to use. The main advantages of `=` were:
+
+-   it is shorter,
+-   it is arguably more natural when defining an implementation, and
+-   we found ourselves reaching for `=` more often in examples.
```

Comment:
I wonder if it's worth mentioning the case of a `where` appearing within an `impl` here. That's one case where I think `=` rather than `==` is a bit less surprising, because it does feel a bit more like you're setting the value of an associated constant.

Reply:
Added https://github.com/carbon-language/carbon-lang/pull/2070/commits/edf74bf49e29a8011b3f5d4ff630f0a31f51da0a